### PR TITLE
[ticket/16908] Fix PHP warning on non-existent post id requests

### PR DESCRIPTION
**Pull request by rxu**
Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

<a href="https://tracker.phpbb.com/browse/PHPBB3-16908">PHPBB3-16908</a>.

### CODE REVIEW

File: phpBB/posting.php [hunk 1]
```diff
@@ -87,24 +87,25 @@
...
 			$result = $db->sql_query($sql);
-			$topic_forum = $db->sql_fetchrow();
-			$topic_id = (int) $topic_forum['topic_id'];
-			$forum_id = (int) $topic_forum['forum_id'];
+			$topic_forum = $db->sql_fetchrow($result);
 			$db->sql_freeresult($result);
 		}
 
-		if (!$post_id || !$topic_id || !$forum_id)
+		if (!$post_id || !$topic_forum)
```
Comment (3D-I) [inline phpBB/posting.php:101]: May I suggest what I wrote to solve this ticket, which is slightly different?

Comment (rxu) [reply phpBB/posting.php:101]: Sure, what code do you have in mind?

Comment (3D-I) [reply phpBB/posting.php:101]: Well, I did delete it. Here is what I wrote now and briefly tested.

```php
	case 'edit':
	case 'delete':
	case 'quote':
	case 'soft_delete':
		$post_id = $request->variable('p', 0);

		$sql = 'SELECT post_id
			FROM ' . POSTS_TABLE . '
			WHERE post_id = ' . $post_id;
		$result = $db->sql_query_limit($sql, 1);
		$row = $db->sql_fetchrow($result);
		$db->sql_freeresult($result);

		$post_id = $row['post_id'] ?? 0;

		if ($post_id)
		{
			$topic_forum = [];

			$sql = 'SELECT t.topic_id, t.forum_id
				FROM ' . TOPICS_TABLE . ' t, ' . POSTS_TABLE . ' p
				WHERE p.post_id = ' . $post_id . '
				AND t.topic_id = p.topic_id';
			$result = $db->sql_query($sql);
			$topic_forum = $db->sql_fetchrow($result);
			$db->sql_freeresult($result);

			$topic_id = (int) $topic_forum['topic_id'];
			$forum_id = (int) $topic_forum['forum_id'];
		}
		else
		{
			$user->setup('posting');
			trigger_error('NO_POST');
		}
	break;
}
```

Comment (rxu) [reply phpBB/posting.php:101]: Well, that additional query actually duplicates the current query (the only difference is the current one joins topics table, but this doesn't change the result).
I'd think about adding a limit of 1 to the query, not sure if it brings any benefits though.

Comment (3D-I) [reply phpBB/posting.php:101]: There is no need to go ahead if the post does not exists and a simple query which asserts it does or not simpliefies the workflow. The same method should also be used above in the same swittch/cases for the forums (IMHO).

Comment (3D-I) [reply phpBB/posting.php:101]: On a side note, is this is a regression due to the same PR which took care of removing the forum queries from links?

Comment (rxu) [reply phpBB/posting.php:101]: Sorry, I'm still on the point those queries do the same thing :)

> is this is a regression due to the same PR which took care of removing the forum queries from links?

Yes, the whole code was added here
https://github.com/phpbb/phpbb/commit/adacf21edb0bdd3e8aaa649ca037586a68ccf5e2#diff-2bd852c3b9a70bfaabb479ef5b67b657056c1e74de09e433e52a644092ec28c9

Comment (3D-I) [reply phpBB/posting.php:101]: Two different points of view, indeed. :)

Comment (3D-I) [reply phpBB/posting.php:101]: Again, ther is no need to run a query which joins two tables if the post ID does not exist to begin with, you can not rely on the request var to assert it.

Comment (rxu) [reply phpBB/posting.php:101]: Post existence is checked with that query. There's no need in additional one. There's no relying on single request value.

**Pull request merged**